### PR TITLE
fix: Save properties after applying changes in Dashboard

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY
This PR fixes a severe issue for which changes in the properties modal of a Dashboard won't be fully saved when the `onlyApply` prop was used. This PR https://github.com/apache/superset/pull/17392 enabled the `onlyApply` tag. Later, manual tests have shown that the tag wasn't ready to be used as both the frontend and the backend were not fully implemented for that to work properly. In addition to that, this PR does the following:

- Refactors the `PropertiesModal` to use typescript
- Refactors the `form` component to use the functionalities provided by Antdesign
- Fixes an issue in the Dashboard detail page for which when applying changes and re-opening the properties modal it did not show the latest changes 
- Fixes an issue for which changes in the `json_metadata` for `color_namespace`, `expanded_slices`, `timed_refresh_immune_slices` were not saved 
- Fixes an issue for which the dashboard wasn't redirecting back to the original URL when the _slug_ was deleted
- Fixes an issue for which changes to native filters would not update the json_metadata immediately, causing potential overwrites
- Changes the endpoint that was used for updating the dashboard (`save_dash`) with the standard PUT endpoint. Also, it adds the `set_dash_metadata` function to the `update` command in the backend. The `set_dash_metadata` function was only used by the `save_dash` endpoint before, causing potential discrepancies in the way the `json_metatada` was handled by the standard PUT endpoint.
- Removes unnecessary fetch requests

### TESTING INSTRUCTIONS
1. Open a Dashboard
2. Edit the properties, including the json metadata
3. Apply the changes
4. Reopen the modal and make sure the latest changes are there
4. Finally, save the changes
5. Reload the Dashboard and make sure the properties were fully saved

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_properties.test.ts
Comment (geido) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_properties.test.ts:100]: This was here because the color scheme was considered to e mandatory. This is not the case any longer

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_properties.test.ts
Comment (geido) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_properties.test.ts:103]: This was here because the color scheme was considered to e mandatory. This is not the case any longer

File: superset-frontend/spec/javascripts/dashboard/components/PropertiesModal_spec.jsx
Comment (geido) [inline superset-frontend/spec/javascripts/dashboard/components/PropertiesModal_spec.jsx:63]: All these tests need to be refactored since this PR changes the component to be functional. I'll work on these in a separate PR to avoid holding this PR too long

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true

**Comment by jinghua-qa** [comment]
I found a bug that edit native filter in meta data did not reflect changes after save. See video for details:

https://user-images.githubusercontent.com/81597121/144770187-2df63a4e-9f46-48f6-a568-b3acb5aa1e24.mov

**Comment by geido** [comment]
Hey @jinghua-qa that seems to be affecting master as well and it is not related to this PR. Would you mind opening a separate issue for that? Thank you!

File: superset/dashboards/commands/update.py
Comment (betodealmeida) [inline superset/dashboards/commands/update.py:58]: If for some reason `json_metadata` is an empty string this would raise an error:

```python
>>> properties = {'json_metadata': ''}
>>> properties.get('json_metadata') is not None
True
>>> json.loads(properties.get("json_metadata", "{}"))
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

You might wanna do instead:

```suggestion
            if self._properties.get("json_metadata"):
                dashboard = DashboardDAO.set_dash_metadata(
                    dashboard,
                    data=json.loads(self._properties.get("json_metadata", "{}")),
```

File: superset/dashboards/dao.py
Comment (betodealmeida) [inline superset/dashboards/dao.py:243]:
```suggestion
        # TODO (geido): remove by refactoring/deprecating save_dash endpoint
```

I'm planning to write a bot that emails committers that leave TODOs open for too long. :)

File: superset/dashboards/dao.py
Comment (betodealmeida) [inline superset/dashboards/dao.py:255]: Python dictionaries have an odd-named method for this:

```suggestion
        md.setdefault("timed_refresh_immune_slices", [])
```

**Review by betodealmeida — COMMENTED** [review]
The backend looks good to me. I'm no longer super familiar with the frontend here, will defer the decision.

File: superset-frontend/cypress-base/cypress/integration/dashboard/save.test.js
Comment (kgabryje) [inline superset-frontend/cypress-base/cypress/integration/dashboard/save.test.js:74]: How on earth did that work 🤦

File: superset-frontend/spec/javascripts/dashboard/components/PropertiesModal_spec.jsx
Comment (kgabryje) [reply superset-frontend/spec/javascripts/dashboard/components/PropertiesModal_spec.jsx:63]: I hate those enzyme tests, great to see them go one by one

File: superset-frontend/src/dashboard/actions/dashboardState.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/dashboardState.js:217]: `owners && owners.length` -> `ensureIsArray(owners)`
`owners.map(o => (o.id ? o.id : o))` - I assume that means that owner can either be a string or an object with id? Maybe we could use a type guard to make it a bit clearer?

File: superset-frontend/src/dashboard/actions/dashboardState.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/dashboardState.js:218]: `isFeatureEnabled(FeatureFlag.DASHBOARD_RBAC) ? ensureIsArray(roles).map(r => (r.id ? r.id : r)) : undefined`

`.map(r => (r.id ? r.id : r))` - same as above (comment about owners)

File: superset-frontend/src/dashboard/actions/dashboardState.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/dashboardState.js:280]: Json metadata is already parsed in line 277 🙂

File: superset-frontend/src/dashboard/actions/dashboardState.js
Comment (kgabryje) [inline superset-frontend/src/dashboard/actions/dashboardState.js:308]: Hmm, I think that in the case of undefined error the message would sound a bit strange: "Sorry, there was an error saving this dashboard: Unknown". Maybe we should have a separate error message for such case? Like "Unknown error has occurred" or something

File: superset-frontend/src/dashboard/components/PropertiesModal/index.tsx
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/PropertiesModal/index.tsx:119]: You can `useCallback` instead and remove the first `() =>`

File: superset-frontend/src/dashboard/components/PropertiesModal/index.tsx
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/PropertiesModal/index.tsx:219]: `ensureIsArray`

File: superset-frontend/src/dashboard/components/PropertiesModal/index.tsx
Comment (kgabryje) [inline superset-frontend/src/dashboard/components/PropertiesModal/index.tsx:433]:
```suggestion
                'Roles is a list which defines access to the dashboard. Granting a role access to a dashboard will bypass dataset level checks. If no roles are defined, then the dashboard is available to all roles.',
```

**Review by kgabryje — CHANGES_REQUESTED** [review]
Great work! I left a few cosmetic comments.

Since this PR is very large, it would be great if more people took a look at the code

**Review by kgabryje — APPROVED** [review]
Thanks for the fixes! LGTM, but like I said, it would be great to have 1 more review since it's a big PR

File: superset-frontend/src/dashboard/components/PropertiesModal/index.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/PropertiesModal/index.tsx:114]: Maybe a `t()` here - not worth holding up the PR over this though. Looks like there are a few more of these floating around in the codebase to sweep up anyway

File: superset-frontend/src/dashboard/components/PropertiesModal/index.tsx
Comment (rusackas) [inline superset-frontend/src/dashboard/components/PropertiesModal/index.tsx:262]:
```suggestion
      throw new Error('A valid color scheme is required');
```
Could use a `t()` here. Not blocking the PR for this!

**Review by rusackas — APPROVED** [review]
LGTM! Thanks for all this hard work!

**Comment by suddjian** [comment]
Thanks for this fix! Embedded work has been keeping me busy so thanks to the others for picking up the review. I managed to finally take a bit of a look today and it LGTM too. The original code here was one of my first Superset contributions 😁

**Pull request merged**
